File: linked_lists.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = self.create_node()
    
    def create_node(self):
        node = {
            "value": self.value,
            "next": None,
        }

        return node


class LinkedLists:
    def __init__(self, value):
        self.value = value
        self.head = {"value": self.value, "next": None}
        self.tail = self.head
        self.length = 1

    def append(self, value):
        node = Node(value)
        end_node = node.next

        self.tail["next"] = end_node
        self.tail = end_node
        self.length += 1

    def prepend(self, value):
        node = Node(value)
        new_node = node.next
        new_node["next"] = self.head
        self.head = new_node
        self.length +=1

    def print_list(self):
        current_node = self.head
        linked_list = []
        while current_node["next"] != None:
            linked_list.append(current_node["value"])
            current_node = current_node["next"]
        linked_list.append(current_node["value"])
        return linked_list
    
    def traverse(self, index):
        current_node = self.head
        counter = 0
        while (counter != index):
            current_node = current_node["next"]
            counter +=1
        return current_node
    
    def remove(self, index):
        lead_node = self.traverse(index-1)
        pointer_node = lead_node["next"]
        new_node = pointer_node["next"]
        lead_node["next"] = new_node
        self.length -= 1
        return self.head

    
    def insert(self, value, index):
        if index >= self.length:
            self.append(value)
        elif index < 1:
            self.prepend(value)
        else:
            node = Node(value)
            new_node = node.next
            lead_node = self.traverse(index-1)
            pre_node = lead_node["next"]
            new_node["next"] = pre_node
            lead_node["next"] = new_node
            self.length += 1
            return self.head
        
    def reverse(self):
        if not self.head["next"]:
            return self.head
        else:
            first = self.head
            self.tail = self.head
            second =  first["next"]

            while second is not None:
                temp  = second["next"]
                second["next"] = first
                first = second
                second = temp

            self.head["next"] = None
            self.head = first
            return self.head

ll = LinkedLists(1)
ll.append(2)
ll.append(3)
# ll.prepend(0)
# print(ll.head)
# print(ll.print_list())
# print(ll.insert(1, 2))
# print(ll.insert(5, 3))
# print(ll.remove(1))
print(ll.reverse())
print(ll.print_list())
print(ll.tail)

File: linkedLists.js
// 10 -> 6 -> 2

class linkedLists {
    constructor(value){
        this.head = {value: value,
            next: null
        };
        this.tail = this.head;
        this.length = 1;
    }

    append(value){
        const newNode = {
            value: value,
            next: null
        }

        this.tail.next = newNode;
        this.tail = newNode;
        this.length ++;
    }

    prepend(value){
        //{v: 2, next: {} }
        const newNode = {
            value: value,
            next: null
        };
        newNode.next = this.head;
        this.head = newNode;
        this.length ++;
    }

    insert(index, value){
        if (index >= this.length){
            this.append(value);
        }
        else {
            const newNode = {
                value: value,
                next: null
            }

            const leadNode = this.traverse(index-1);
            const pointerNode = leadNode.next;
            newNode.next = pointerNode;
            leadNode.next = newNode;
            this.length ++;
            return this.head;
        }
    }

    delete(index){
        const leadNode = this.traverse(index-1);
        const pointerNode = leadNode.next.next;
        leadNode.next = pointerNode;
        this.length -- ;
        return this.head;
    }

    traverse(index){
        let currentNode = this.head;
        let i = 0;
        while (i < index){
            currentNode = currentNode.next;
            i ++;
        }

        return currentNode;
    }



    printList(){
        let linkList = [];
        let currentNode = this.head.next;
        linkList.push(this.head.value);
        while (currentNode.next != null){
            linkList.push(currentNode.value);
            currentNode = currentNode.next;
        }
        linkList.push(currentNode.value);
        return linkList;
    }

    reverse(){
        if (!this.head.next){
            return this.head
        }
        else {
            let first = this.head;
            this.tail = this.head;
            let second = first.next; //this.head.next
            while (second){
                let temp = second.next;
                second.next = first; //  {this.head}
                first = second;
                second = temp;
            }
            this.head.next = null;
            this.head = first;
            return this.head;
        }
    }
}

const ll = new linkedLists(10);
// console.log(ll);
ll.append(6);
ll.append(2);
ll.prepend(1);
console.log(ll.insert(2,2));
//console.log(ll.insert(2,3));
console.log(ll.printList());
console.log(ll.delete(2));
console.log(ll.printList());
// console.log(ll.head.next);
// console.log(ll.list);
// console.log(ll.insert(3, 10));
// console.log(ll.list);
console.log(ll.reverse())
console.log(ll.printList());